### PR TITLE
Implement handling of framework-handled key events

### PR DESCRIPTION
**Pull request by gspencergoog**
## Description

This implements the "Delayed event delivery" that I've implemented for the other platforms.

Because they require synchronous responses, other platforms need to re-dispatch events that weren't handled by the framework once it responds, but the web just needs to call `preventDefault` on the events that the framework handles, since it all happens in the same thread anyhow.

## Related Issues

- flutter/flutter#47156

## Tests

- Added tests to check that events handled by the framework have `preventDefault` called on them (and vice versa).

## Breaking Change

- [X] No, no existing tests failed, so this is *not* a breaking change.

### CODE REVIEW

File: lib/web_ui/lib/src/engine/keyboard.dart [hunk 1]
```diff
@@ -132,16 +127,18 @@ class Keyboard {
...
+          return;
+        }
+        final String response = utf8.decode(data.buffer.asUint8List());
+        final Map<String, dynamic> jsonResponse = json.decode(response);
+        if (jsonResponse['handled'] as bool) {
+          // If the framework handled it, then don't propagate it any further.
+          event.preventDefault();
+        }
+      },
+    );
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/keyboard.dart:141]: FYI this file is being heavily refactored in https://github.com/flutter/engine/pull/23466.

Comment (gspencergoog) [reply lib/web_ui/lib/src/engine/keyboard.dart:141]: Yeah, I'm quite aware (and Tong knows I'm working on this).

**Comment by gspencergoog** [comment]
@mdebbar OK, this is ready for review, if you have the time.

File: lib/web_ui/lib/src/engine/keyboard.dart [hunk 2]
```diff
@@ -132,16 +127,18 @@ class Keyboard {
...
-        return true;
-      default:
-        return false;
-    }
+      _messageCodec.encodeMessage(eventData), (ByteData? data) {
+        if (data == null) {
+          return;
+        }
+        final String response = utf8.decode(data.buffer.asUint8List());
+        final Map<String, dynamic> jsonResponse = json.decode(response);
```
Comment (mdebbar) [inline lib/web_ui/lib/src/engine/keyboard.dart:135]: I'm curious why not use a `JSONMessageCodec` to decode `data`? There's already one called `_messageCodec` in this file.

File: lib/web_ui/test/keyboard_test.dart [hunk 3]
```diff
@@ -226,13 +235,23 @@ void testMain() {
...
 
-    test('prevents default when "Tab" is pressed', () {
+    test('prevents default when key is handled by the framework', () {
+      ByteData _toByteData(List<int> bytes) {
+        final ByteData byteData = ByteData(bytes.length);
+        for (int i = 0; i < bytes.length; i++) {
+          byteData.setUint8(i, bytes[i]);
+        }
+        return byteData;
+      }
```
Comment (mdebbar) [inline lib/web_ui/test/keyboard_test.dart:245]: Is it possible to avoid this if you use `JSONMessageCodec.encodeMessage()`?

**Review by mdebbar — APPROVED** [review]

File: lib/web_ui/lib/src/engine/keyboard.dart (hunk 2 above)
Comment (gspencergoog) [reply lib/web_ui/lib/src/engine/keyboard.dart]: No good reason, I'm just a goofball.  Converted to use `_messageCodec`, which makes a lot more sense.

File: lib/web_ui/test/keyboard_test.dart (hunk 3 above)
Comment (gspencergoog) [reply lib/web_ui/test/keyboard_test.dart]: Yes.  Done.

**Pull request merged**